### ISSUE TITLE
Trigger onChange after schema/normalize

### ISSUE THREAD

**Issue by Mangatt** [ianstormtaylor/slate#575]
Hi, when plugin implements schema rules and normalize by them, it works fine, but these changes are not sent to state. This effectively mean that saved and displayed states are out of sync.

So I guess there should be onChange event triggered after normalizing.

**Comment by ianstormtaylor** [comment]
Hey @Mangatt can you describe your use case where this is causing issues?

I agree with you at first glance, it seems like plugins should be able to access the "true" state of the editor, instead of only the pre-normalized state. 

I initially thought we could move the normalize into the `onBeforeChange` stack, but right now the way the accumulator handlers like `onBeforeChange` and `onChange` work, this wouldn't be possible, since they can always return a new value, and normalization has to occur after any place where a new value can be returned.

But there's also a possibility that we could add `onAfter*` handlers, which are purely observers, which might solve your case. But I'd like to know more first so that I can be sure it's the right solution! Thanks

**Comment by Mangatt** [comment]
My problem first occured with lists, when I wrapped selected items in list and let rules handle wrapping blocks in list in LI's.

But there are some more problems with normalizing with schema:
- toolbar buttons do not necessarily show correct state of selection (block type is selected even when rule already changed this block into something else)
- before saving to server, you need to trigger focus / blur to ensure that showed state will be really the one that will be saved
- findDOMNode is not reliable since nodes can change on the fly

What I am proposing is that when state is normalized by any of the plugins, there should be triggered standard onChange event by slate, after all plugins are done with normalizing. Just to flush changes to parent component, to ensure internal and external states are same.

**Comment by ianstormtaylor** [comment]
That should already be happening though—that the parent `Editor` will re-receive new `props.state` with the newly-changed, fully-normalized `state`?

**Comment by Mangatt** [comment]
No, it's not happening right now. I'm managing state with redux and plugin rule is merging sibling lists into one.

Editor shows only merged list, but redux has state before merging (multiple lists). After I focus editor, correct state is sent to redux.

**Comment by ianstormtaylor** [comment]
Can you make a very simple test case that reproduces this so it's easier to debug?

**Comment by Mangatt** [comment]
That might be difficult since you have to track saved state and I think that jsfiddle does not support that. Here is my rule that joins sibling lists:

```javascript
{
	match: node => node.kind == 'block' || node.kind == 'document',
	validate(block) {
		//Join sibling lists
		const lists = block.nodes.filter((block, key, blocks) => block.type == type && key && blocks.get(key-1).type == type);
		return lists.size ? lists : null;
	},
	normalize(transform, block, lists) {
		lists.map(list => {
			let prev = transform.state.document.getPreviousSibling(list.key);
			list.nodes.map(listItem => {
				transform.moveNodeByKey(listItem.key, prev.key, Infinity);
			});
			transform.removeNodeByKey(list.key);
		})
		return transform;
	}
}
```

Can you use this to reproduce that problem? If not, I can try to think something out at jsfiddle eventually.

**Comment by YurkaninRyan** [comment]
@Mangatt Did you happen to find a solution to this problem?  I'm having a similar problem that i'm trying to recreate in a fiddle, but no matter what I do it works perfectly in the fiddle (checked React Versions, bare bones in my app, etc... )

**Comment by YurkaninRyan** [comment]
Fixed half of it today by upgrading from `0.20.2` to `0.20.4`.  It seems like the normalize calls are still not firing an onChange event when the normalize happens on initial load.  e.g if I have a rule where there must always be one PARAGRAPH element.

1. Deserialize Empty Html
2. Get Empty State, pass into editor.
3. Normalize happens to add PARAGRAPH, but editor doesn't update until I focus.

My quick fix for this was just to auto focus the editor :+1:

**Comment by ianstormtaylor** [comment]
This issue hasn't had a response in a while, and it's unclear if it's still present in the latest versions of Slate. To keep the issues easier to manage, since there are so many being opened every day, I'm going to close this old one out. 

But if you notice that the bug is still happening, please feel free to comment and it can be revisited. It might be an edge case that not many people run into, in which case the fastest way to get it fixed would probably be to write a pull request for it.

Thanks for understanding!
